### PR TITLE
added linelint workflow

### LINKED ISSUE #3519: Create a workflow detect EOF and fix it for PR automatically
### Summary

Seems linelint[1] can help us to do this.

### Appendix

1. https://github.com/marketplace/actions/linelint

### PR DESCRIPTION
**Pull request by arujjval**
closes #3519 

#### Notes
Previous discussions about this PR were in #3558 #3559 #3560

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by dannypsnl** [comment]
In the log we can see,
```
[EOF Rule] File ".clang-tidy" has lint errors
[EOF Rule] File ".clang-tidy" lint errors fixed
```
However, there are no changes are proposed to GitHub, you need to fill that gap.

**Comment by arujjval** [comment]
@dannypsnl This is fixing eof errors in our repo as we have made a PR. This is what was requested in the issue #3592.

**Comment by hydai** [comment]
@dannypsnl 
I changed my mind. Let's use this linelint workflow only for detection. I don't think using autofix is a good idea now if a malicious hacker hijacks the linelint workflow, and it will allow the workflow to inject or modify some code into our codebase directly.

@0yi0 WDYT?

File: .github/workflows/linelint.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+jobs:
+  linelint:
+    runs-on: ubuntu-latest
+    name: Check if all files end in newline
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Linelint
+        uses: fernandrone/linelint@master
```
Comment (0yi0) [inline .github/workflows/linelint.yml:12]: Already mentioned in the old PRs, please use a tag instead of master.

**Comment by hydai** [comment]
We should also add the following limitation to the workflow file to keep it safe:

```
permissions:
  contents: read
```

File: .github/workflows/linelint.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+on: pull_request
+name: linelint
+
+jobs:
+  linelint:
+    runs-on: ubuntu-latest
+    name: Check if all files end in newline
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (0yi0) [inline .github/workflows/linelint.yml:10]: `actions/checkout@v2` is a very legacy version, please use `v4` or add a reason for it.

**Review by 0yi0 — CHANGES_REQUESTED** [review]

**Comment by 0yi0** [comment]
@hydai No autofix sounds good to me. But if we're not running autofix, should CI report failure on lint errors?

**Comment by hydai** [comment]
Definitely, we should let CI fail if the changed files do not contain EOF. Just like what we did in the misc linter.

**Comment by 0yi0** [comment]
@arujjval dannypsnl means your workflows only fix it in the workflows **without** any modifications to the code base (the PR or the reporitory). So `autofix` being `true` or `false` didn't matter in the PR because no changes were committed.

**Comment by arujjval** [comment]
So I have to remove autofix, and only check incoming changes for EOF error?

**Comment by 0yi0** [comment]
@arujjval I'm wondering if we can keep `autofix: true` in `.linelint`, and use `autofix: false` in `linelint.yml`, so that when `linelint.yml` reports error, user can use your `.linelint` config directly to fix their issues.

If this works, `.linelint` is already good and need no modifications. You only need to update `linelint.yml`, and then we can merge.

I hope I explained it well. Please let met know if you have questions about this.

**Comment by arujjval** [comment]
I am not able to make this lint run only on changed modified files. Instead of running a command, I am "using" the repo of linelint. One thing we can do is install linelint in the repo and then running script in workflow. Can we that or is there any other way round?

**Comment by dannypsnl** [comment]
> I am not able to make this lint run only on changed modified files.

I believe you can add
```sh
git diff <branch> --name-status
```
for example, this PR merge 2 commits into master, so
```sh
git diff master --name-status
```
let sed out the files.

**Comment by arujjval** [comment]
I am able to get modified files and I am passing it while running linelint, but it is still running on all files. What can be the potential issues?
Screenshot from 2024-07-30 12-14-47

**Comment by 0yi0** [comment]
I checked the source code of this action, it doesn't seem to provide input fields.

It runs `linelint` like this
```
ENTRYPOINT ["/linelint", "."]
```

It's probably the same as the warning at the first line in your screenshot.

> Unexpected input(s) 'files', valid inputs are ['entryPoint', 'args']

I'm not sure if that means you can change `['entryPoint', 'args']` from input though.
If yes, you can try to set `args` to the list of files.

**Comment by arujjval** [comment]
@0yi0 turns out we can change `args` from input. And it seems to work. But the issue is it is not ignoring other files. Here's the result after running with args and without args

Screenshot from 2024-07-30 16-26-12
with args

Screenshot from 2024-07-30 16-25-41
without args

`args` seems to add files for lint check but it does not exclude other files. Or these files might be special case. Maybe we can correct lint error of these files (they are just 4) and then try.

**Comment by hydai** [comment]
Hi @arujjval 
I'm sorry I missed this PR. Could you remind me of its current state? Is there any help you need to continue working on it?

**Comment by hydai** [comment]
Closing due to inactive. Thanks.

**Pull request closed without merging**